### ISSUE TITLE
Regression in 0.44 or 0.45: Reconciler future is no longer started immediately

### ISSUE THREAD

**Issue by joell** [clux/kube-rs#372]
Somewhere between 0.43 and 0.45 the output `Stream` of `kube_runtime::controller::Controller::run` began exhibiting a delay between when a reconciliation `Err` result was returned from the reconciliation function and when it actually appeared in the stream.  I've diagnosed this a bit and it appears the `Err` result is not emitted to the stream until after the **next** IO action takes place in the Controller; in cases where one is watching a rarely changing set of CRD resources, this could result in a very long delay before the error is reported.

This problem is also present in 0.46.

To reproduce the issue, here is a slightly modified variant of the `ConfigMapGenerator` example included in the documentation:
```rust
use futures::StreamExt;
use k8s_openapi::api::core::v1::ConfigMap;
use kube::{Client, api::{Api, ListParams}};
use kube_derive::CustomResource;
use kube_runtime::controller::{Context, Controller, ReconcilerAction};
use log::{error, info};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use snafu::Snafu;
use tokio::time::Duration;


#[derive(Debug, Snafu)]
enum Error { ExampleFailure }
/// A custom resource
#[derive(CustomResource, Debug, Clone, Deserialize, Serialize, JsonSchema)]
#[kube(group = "nullable.se", version = "v1", kind = "ConfigMapGenerator", namespaced)]
struct ConfigMapGeneratorSpec {
    content: String,
}

/// The reconciler that will be called when either object change
async fn reconcile(g: ConfigMapGenerator, _ctx: Context<()>) -> Result<ReconcilerAction, Error> {
    info!("ConfigMapGenerator to reconcile: {:?}", g);
    Err(Error::ExampleFailure)
}
/// an error handler that will be called when the reconciler fails
fn error_policy(_error: &Error, _ctx: Context<()>) -> ReconcilerAction {
    ReconcilerAction {
        requeue_after: Some(Duration::from_secs(60)),
    }
}

/// something to drive the controller
#[tokio::main]
async fn main() -> Result<(), kube::Error> {
    pretty_env_logger::init_timed();

    let client = Client::try_default().await?;
    let context = Context::new(()); // bad empty context - put client in here
    let cmgs = Api::<ConfigMapGenerator>::all(client.clone());
    // let cms = Api::<ConfigMap>::all(client.clone());
    Controller::new(cmgs, ListParams::default())
        // .owns(cms, ListParams::default())   // <-- NEED TO IMMEDIATELY SHOW ERROR
        .run(reconcile, error_policy, context)
        .for_each(|res| async move {
            match res {
                Ok(o)  => info!("reconciled {:?}", o),
                Err(e) => error!("reconcile failed: {:?}", e),
            }
        })
        .await; // controller does nothing unless polled
    Ok(())
}
```

The major changes I've made are:

* Changed the `reconcile` function to always fail.
* Added a logging subsystem (the `pretty_env_logger` and `log` crates), so we can see debug hyper logs to determine when HTTP IO is happening.
* Made the `reconcile` function log its input, so we can tell when it was called.
* Commented out the `Controller::owns` method call, so only `ConfigMapGenerator` resources trigger reconciliation.

(I retained the same `ConfigMapGenerator` YAML definition as you have in your `examples/` directory.)

When I run this code with the `ConfigMapGenerator` CRD defined in my K8S cluster and a single custom resource of that type called `foo` present in the cluster, I can see the reconcile function triggers, but the error is not consistently logged.  However, if I then patch a change into the `foo` custom resource, this triggers more HTTP IO on the watch and **then** error is reported, as well the reconcile function being triggered again.

Here is console output of my performing that test (with a note midway of when I did the patch):
```
$ RUST_LOG=debug cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running `target/debug/example`
 2021-01-05T17:05:08.919Z DEBUG kube::config > failed to load client identity from kubeconfig: Error loading kubeconfig: Failed to get data/file with base64 format
 2021-01-05T17:05:09.033Z DEBUG reqwest::connect > starting new connection: https://localhost:6443/
 2021-01-05T17:05:09.033Z DEBUG hyper::client::connect::dns > resolving host="localhost" 
 2021-01-05T17:05:09.033Z DEBUG hyper::client::connect::http > connecting to 127.0.0.1:6443 
 2021-01-05T17:05:09.034Z DEBUG hyper::client::connect::http > connected to 127.0.0.1:6443 
 2021-01-05T17:05:09.036Z DEBUG hyper::proto::h1::io         > flushed 191 bytes 
 2021-01-05T17:05:09.037Z DEBUG hyper::proto::h1::io         > read 1186 bytes 
 2021-01-05T17:05:09.037Z DEBUG hyper::proto::h1::io         > parsed 4 headers 
 2021-01-05T17:05:09.037Z DEBUG hyper::proto::h1::conn       > incoming body is content-length (1224 bytes) 
 2021-01-05T17:05:09.038Z DEBUG reqwest::async_impl::client  > response '200 OK' for https://localhost:6443/apis/nullable.se/v1/configmapgenerators?
 2021-01-05T17:05:09.038Z DEBUG hyper::proto::h1::io         > read 182 bytes 
 2021-01-05T17:05:09.038Z DEBUG hyper::proto::h1::conn       > incoming body completed 
 2021-01-05T17:05:09.038Z DEBUG hyper::client::pool          > pooling idle connection for ("https", localhost:6443) 
 2021-01-05T17:05:09.039Z DEBUG hyper::client::pool          > reuse idle connection for ("https", localhost:6443) 
 2021-01-05T17:05:09.039Z DEBUG hyper::proto::h1::io         > flushed 244 bytes 
 2021-01-05T17:05:09.040Z DEBUG hyper::proto::h1::io         > read 150 bytes 
 2021-01-05T17:05:09.040Z DEBUG hyper::proto::h1::io         > parsed 4 headers 
 2021-01-05T17:05:09.040Z DEBUG hyper::proto::h1::conn       > incoming body is chunked encoding 
 2021-01-05T17:05:09.040Z DEBUG reqwest::async_impl::client  > response '200 OK' for https://localhost:6443/apis/nullable.se/v1/configmapgenerators?&watch=true&resourceVersion=294708&timeoutSeconds=290
 2021-01-05T17:05:09.041Z INFO  example                      > ConfigMapGenerator to reconcile: ConfigMapGenerator { api_version: "nullable.se/v1", kind: "ConfigMapGenerator", metadata: ObjectMeta { annotations: Some({"kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"nullable.se/v1\",\"kind\":\"ConfigMapGenerator\",\"metadata\":{\"annotations\":{},\"name\":\"foo\",\"namespace\":\"example\"},\"spec\":{\"content\":\"Hey!\"}}\n"}), cluster_name: None, creation_timestamp: Some(Time(2021-01-05T17:02:40Z)), deletion_grace_period_seconds: None, deletion_timestamp: None, finalizers: None, generate_name: None, generation: Some(5), labels: None, managed_fields: Some([ManagedFieldsEntry { api_version: Some("nullable.se/v1"), fields_type: Some("FieldsV1"), fields_v1: Some(FieldsV1(Object({"f:metadata": Object({"f:annotations": Object({".": Object({}), "f:kubectl.kubernetes.io/last-applied-configuration": Object({})})}), "f:spec": Object({})}))), manager: Some("kubectl-client-side-apply"), operation: Some("Update"), time: Some(Time(2021-01-05T17:02:40Z)) }, ManagedFieldsEntry { api_version: Some("nullable.se/v1"), fields_type: Some("FieldsV1"), fields_v1: Some(FieldsV1(Object({"f:spec": Object({"f:content": Object({})})}))), manager: Some("kubectl-patch"), operation: Some("Update"), time: Some(Time(2021-01-05T17:05:04Z)) }]), name: Some("foo"), namespace: Some("example"), owner_references: None, resource_version: Some("294706"), self_link: Some("/apis/nullable.se/v1/namespaces/example/configmapgenerators/foo"), uid: Some("393c181b-239d-40fd-9641-8a9893151207") }, spec: ConfigMapGeneratorSpec { content: "1" } }


<<< Here is when I did the "kubectl patch" on the "foo" resource to trigger new watch IO >>>


 2021-01-05T17:05:12.806Z DEBUG hyper::proto::h1::io         > read 1078 bytes 
 2021-01-05T17:05:12.806Z DEBUG hyper::proto::h1::decode     > incoming chunked header: 0x42F (1071 bytes) 
 2021-01-05T17:05:12.806Z ERROR example                      > reconcile failed: ReconcilerFailed { source: ExampleFailure, backtrace: Backtrace(()) }
 2021-01-05T17:05:12.811Z INFO  example                      > ConfigMapGenerator to reconcile: ConfigMapGenerator { api_version: "nullable.se/v1", kind: "ConfigMapGenerator", metadata: ObjectMeta { annotations: Some({"kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"nullable.se/v1\",\"kind\":\"ConfigMapGenerator\",\"metadata\":{\"annotations\":{},\"name\":\"foo\",\"namespace\":\"example\"},\"spec\":{\"content\":\"Hey!\"}}\n"}), cluster_name: None, creation_timestamp: Some(Time(2021-01-05T17:02:40Z)), deletion_grace_period_seconds: None, deletion_timestamp: None, finalizers: None, generate_name: None, generation: Some(6), labels: None, managed_fields: Some([ManagedFieldsEntry { api_version: Some("nullable.se/v1"), fields_type: Some("FieldsV1"), fields_v1: Some(FieldsV1(Object({"f:metadata": Object({"f:annotations": Object({".": Object({}), "f:kubectl.kubernetes.io/last-applied-configuration": Object({})})}), "f:spec": Object({})}))), manager: Some("kubectl-client-side-apply"), operation: Some("Update"), time: Some(Time(2021-01-05T17:02:40Z)) }, ManagedFieldsEntry { api_version: Some("nullable.se/v1"), fields_type: Some("FieldsV1"), fields_v1: Some(FieldsV1(Object({"f:spec": Object({"f:content": Object({})})}))), manager: Some("kubectl-patch"), operation: Some("Update"), time: Some(Time(2021-01-05T17:05:12Z)) }]), name: Some("foo"), namespace: Some("example"), owner_references: None, resource_version: Some("294712"), self_link: Some("/apis/nullable.se/v1/namespaces/example/configmapgenerators/foo"), uid: Some("393c181b-239d-40fd-9641-8a9893151207") }, spec: ConfigMapGeneratorSpec { content: "2" } }
```

Notice that the first error is not logged until **after** the patch takes place, even though the `INFO` log line from the reconcile function shows it triggered on the `foo` resource immediately on startup.

If you uncomment the `Controller::owns` method call, sometimes the initial error is immediately reported (but sometimes it may still not be).  This seems to be due to the extra HTTP IO being done to track the generated `ConfigMap` resources, and when its IO falls after the reconcile failure, it can trigger the reconcile error being pushed to the output Stream promptly.

In summary, a change introduced in 0.44 or 0.45 appears to have deferred pushing reconcile errors to the `Controller::run` output stream until after the next bit of HTTP IO.

If there is any further information I can provide that might assist in figuring out why this is happening, please let me know.

Thank you for this project and all the work and time you've put into it! :smiley:

**Comment by teozkr** [comment]
Okay, I am able to reproduce the problem..

**Comment by teozkr** [comment]
Looks like `FutureHashMap` isn't getting polled again once the reconciler returns.. are we losing the `Waker` somewhere along the line?
